src/mds/MDSDmclockScheduler.cc

```diff
@@ -0,0 +1,900 @@
+#include "MDSDmclockScheduler.h"
+#include "mds/MDSMap.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << "mds." << get_nodeid() << ".dmclock_scheduler "
+
+using std::string;
+using TOPNSPC::common::cmd_getval;
+
+ostream& operator<<(ostream& os, VolumeInfo* vi)
+{
+  if (vi) {
+    os << "VolumeInfo: (session_cnt " << vi->get_session_cnt() << ") "
+        << " reservation " << vi->get_reservation()
+        << " weight" << vi->get_weight()
+        << " limit " << vi->get_limit();
+  } else {
+    os << "VolumeInfo has nullptr";
+  }
+  return os;
+}
+
+const VolumeId MDSDmclockScheduler::convert_subvol_root(const VolumeId &volume_id)
+{
+  filepath subvol_root_path(volume_id);
+
+  if (subvol_root_path.depth() > SUBVOL_ROOT_DEPTH) {
+    return "/" + subvol_root_path.prefixpath(SUBVOL_ROOT_DEPTH).get_path();
+  }
+
+  return volume_id;
+}
+
+const VolumeId MDSDmclockScheduler::get_volume_id(Session *session)
+{
+  ceph_assert(session != nullptr);
+  auto client_root_entry = session->info.client_metadata.find("root");
+  if (client_root_entry == session->info.client_metadata.end() || client_root_entry->second == "") {
+    return "";
+  }
+  return convert_subvol_root(client_root_entry->second);
+}
+
+const VolumeId MDSDmclockScheduler::get_session_id(Session *session)
+{
+  ceph_assert(session != nullptr);
+  return std::to_string(session->info.inst.name.num());
+}
+
+template<typename R>
+void MDSDmclockScheduler::enqueue_client_request(const R &mds_req, VolumeId volume_id, Cost cost)
+{
+  dout(10) << __func__ << " volume_id " << volume_id << dendl;
+
+  std::unique_lock<std::mutex> lock(queue_mutex);
+  request_queue.emplace_back(new ClientRequest(mds_req, volume_id, crimson::dmclock::get_time(), cost));
+  /* wake up*/
+  lock.unlock();
+  queue_cvar.notify_all();
+}
+
+void MDSDmclockScheduler::handle_mds_request(const MDSReqRef &mds_req)
+{
+  ceph_assert(mds_is_locked_by_me());
+  dout(10) << __func__ << " " << *mds_req << dendl;
+
+  if (mds->mds_dmclock_scheduler->default_conf.is_enabled() == true
+      && mds_req->get_orig_source().is_client() &&
+      mds->is_active() &&
+      check_volume_info_updated(get_volume_id(mds->get_session(mds_req))) &&
+      check_volume_info_validity(get_volume_id(mds->get_session(mds_req)))
+      ) {
+    Cost cost = get_op_cost(mds_req->get_op());
+    enqueue_client_request<MDSReqRef>(mds_req, get_volume_id(mds->get_session(mds_req)), cost);
+  } else {
+    mds->server->handle_client_request(mds_req);
+  }
+}
+
+void MDSDmclockScheduler::submit_request_to_mds(const VolumeId& vid, std::unique_ptr<ClientRequest>&& request,
+                                                const PhaseType& phase_type, const uint64_t cost)
+{
+  dout(10) << __func__ << " volume_id " << vid << dendl;
+
+  hit_volume_throttle(vid, request->time);
+
+  const MDSReqRef& req = request->mds_req_ref;
+
+  ceph_assert(!mds_is_locked_by_me());
```

Comment:
When building with RelWithDebInfo, mds_is_locked_by_me() always returns true and ceph_assert() takes effect.

it would be changed to that:
```
ceph_assert(!ceph::mutex_debuging || !mds_is_locked_by_me());
```